src/kernels/kernelProvider.base.ts

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+import type { KernelMessage } from '@jupyterlab/services';
+import { Event, EventEmitter, NotebookDocument, Uri, workspace } from 'vscode';
+import { IVSCodeNotebook } from '../platform/common/application/types';
+import { traceInfoIfCI, traceVerbose, traceWarning } from '../platform/logging';
+import { getDisplayPath } from '../platform/common/platform/fs-paths';
+import { IAsyncDisposable, IAsyncDisposableRegistry, IDisposableRegistry } from '../platform/common/types';
+import { noop } from '../platform/common/utils/misc';
+import { IKernel, IKernelProvider, KernelOptions } from './types';
+
+export abstract class BaseKernelProvider implements IKernelProvider {
```

Comment:
KernelProvider creates the Kernel class directly so I had to split it into two parts as well. This is the base class for them both.